File: 2025/Day4/main.py
import numpy

SAMPLE_DATA: bool = False
if SAMPLE_DATA:
    filename = "sample_data.txt"
else:
    filename = "data.txt"


with open(filename, 'r') as f:
    data = f.readlines()


def count_adjacent_rolls(data: numpy.ndarray, pos: tuple[int, int]) -> bool:
    r, c = pos
    v = numpy.count_nonzero(data[max(0, r-1):r+2, max(0, c-1):c+2])
    return (v - 1) < 4

grid = numpy.array([list(map(lambda x: 0 if x == '.' else 1, line.strip())) for line in data], dtype=numpy.uint8)
grid_copy = grid.copy()
p1 = 0
p2 = 0
turn = 0
while True:
    found = False
    rows, cols = grid.shape
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == 1:
                adj = count_adjacent_rolls(grid, (r, c))
                if turn == 0:
                    p1 += adj
                p2 += adj
                if adj:
                    found = True
                    grid_copy[r][c] = 0
    turn += 1
    if not found:
        break
    grid = grid_copy.copy()
print(f'{p1} {p2}')

File: 2025/Day4/main.cpp

#include <algorithm>
#include <charconv>
#include <fstream>
#include <ranges>
#include <string>
#include <string_view>
#include <vector>
#include <cstdio>
#include <unordered_map>
#include <numeric>
#include <print>

namespace r = std::ranges;
namespace v = std::ranges::views;
using namespace std::string_view_literals;

constexpr bool SAMPLE_DATA = false;

std::vector<std::string> read_file(std::string_view filename) {
    std::ifstream file(filename.data());
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(file, line)) {
        lines.push_back(line);
    }
    return lines;
}

using mat = std::vector<std::vector<uint8_t>>;

bool count_adjacent_rolls(const mat& data, std::pair<int64_t, int64_t> pos) {
    auto [r, c] = pos;
    int v = 0;
    for (int i = std::max(0ll, r - 1); i <= std::min<int64_t>(data.size() - 1, r + 1); ++i) {
        for (int j = std::max(0ll, c - 1); j <= std::min<int64_t>(data[0].size() - 1, c + 1); ++j) {
            v += data[i][j];
        }
    }
    return (v - 1) < 4;
}

int main() {
    auto lines = read_file(SAMPLE_DATA ? "sample_data.txt"sv : "data.txt"sv);
    mat grid;
    for (const auto& line : lines) {
        std::vector<uint8_t> row;
        for (char c : line) {
            row.push_back(c == '@' ? 1 : 0);
        }
        grid.push_back(std::move(row));
    }
    int64_t p1 = 0;
    int64_t p2 = 0;
    int turn = 0;
    while (true) {
        bool found = false;
        mat grid_copy = grid;
        for (int64_t r = 0; r < grid.size(); ++r) {
            for (int64_t c = 0; c < grid[0].size(); ++c) {
                if (grid[r][c] == 1) {
                    bool adj = count_adjacent_rolls(grid, {r, c});
                    if (turn == 0) {
                        p1 += adj;
                    }
                    p2 += adj;
                    if (adj) {
                        found = true;
                        grid_copy[r][c] = 0;
                    }
                }
            }
        }
        if (!found) {
            break;
        }
        grid = std::move(grid_copy);
        ++turn;
    }
    std::println("{} {}", p1, p2);
}